packages/react-server/core/renderMiddleware.js

```diff
@@ -105,16 +105,25 @@ module.exports = function(server, routes) {
 			navigateDfd.resolve();
 
 			if (err) {
-				logger.log("onNavigate received a non-2xx HTTP code", err);
-				handleResponseComplete(req, res, context, start, page);
-				if (err.status && err.status === 404) {
-					next();
-				} else if (err.status === 301 || err.status === 302 || err.status === 307) {
+				// The page can elect to proceed to render
+				// even with a non-2xx response.  If it
+				// _doesn't_ do so then we're done.
+				var done = !(page && page.getHaveDocument());
+
+				if (err.status === 301 || err.status === 302 || err.status === 307) {
 					res.redirect(err.status, err.redirectUrl);
-				} else {
-					next(err);
+				} else if (done) {
```

Comment:
Ugh, this diff wound up looking weird...  Moved some things around in here, but the behavior should still be appropriate:

- If the page _hasn't_ advertised `hasDocument` then we're done _no matter what `err` is_.  301, 500, whatever.
- If the page _has_ advertised `hasDocument` then we'll proceed to render.

> 301, 302 aren't supposed to have additional content?

It's valid (and recommended) to provide a small body for redirects.

https://tools.ietf.org/html/rfc2616#section-10.3.2

> > Unless the request method was HEAD, the entity of the response SHOULD contain a short hypertext note with a hyperlink to the new URI(s).